### PR TITLE
Show all Deref implementations recursively

### LINKED ISSUE #87783: Rustdoc no longer shows 'Methods from Deref impls' at all
I tried this code:

```rust
    use std::path::{Path, PathBuf};

    pub struct Foo(PathBuf);
    impl std::ops::Deref for Foo {
        type Target = Path;
        fn deref(&self) -> &Path { &self.0 }
    }
```

I expected to see this happen: Rustdoc shows the docs for `Path::from_str` on the docs for `Foo`, like it used to in 1.51: 
image

Instead, this happened: Rustdoc does not generate those methods:
image

### Meta

searched nightlies: from nightly-2021-04-01 to nightly-2021-07-30
regressed nightly: nightly-2021-06-16
searched commits: from https://github.com/rust-lang/rust/commit/539d7bd3998d9bfed14c264eacda30097a4ea768 to https://github.com/rust-lang/rust/commit/607d6b00d4e0e0475b8de9d0c870b7126fdcdf6b
regressed commit: https://github.com/rust-lang/rust/commit/d74b36ea2f814b720c39d7b60aecaefe512a056b

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.0</summary>

Host triple: x86_64-unknown-linux-gnu
Reproduce with:
```bash
cargo bisect-rustc --script ./docs-exist.sh --start 2021-04-01 
```

where `docs-exist.sh` is
```sh
#!/bin/sh
cargo doc
grep 'Methods from Deref' ${CARGO_TARGET_DIR:-target}/doc/hello_world/foo/struct.Foo.html
```

</details>

cc @pnkfelix

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #87783.

This is a re-implementation of #80653, so taking the original PR comment:

This changes `rustdoc` to recursively follow `Deref` targets so that methods from all levels are added to the rendered output. This implementation displays the methods from all levels in the expanded state with separate sections for each level.

image

cc @camelid
r? @jyn514

### CODE REVIEW

**Review by jyn514 — CHANGES_REQUESTED** [review]
@GuillaumeGomez can you also test whether this fixes https://github.com/rust-lang/rust/issues/84630?

My understanding is that this doesn't re-introduce https://github.com/rust-lang/rust/issues/82465, since the test `src/test/rustdoc/issue-82465-asref-for-and-of-local.rs` hasn't changed - is that correct?

File: src/test/rustdoc/recursive-deref.rs [hunk 1]
```diff
@@ -13,8 +20,96 @@ impl Deref for A {
...
+impl Deref for E {
+    type Target = F;
+
+    fn deref(&self) -> &Self::Target {
+        panic!()
+    }
+}
+
+// @has recursive_deref/struct.F.html '//h3[@class="code-header in-band"]' 'impl Deref for F'
+// @!has '-' '//*[@id="deref-methods-G"]'
```
[the remark this answers is not in the document]
Comment (jyn514) [reply src/test/rustdoc/recursive-deref.rs:73]: Gotcha. Can you add that as a comment?

File: src/test/rustdoc/deref-recursive.rs [hunk 2]
```diff
@@ -0,0 +1,40 @@
+// #26207: Show all methods reachable via Deref impls, recursing through multiple dereferencing
+// levels if needed.
```
[the remark this answers is not in the document]
Comment (jyn514) [reply src/test/rustdoc/deref-recursive.rs:2]: That makes sense :+1: can you add it as a comment?

File: src/librustdoc/passes/collect_trait_impls.rs [hunk 3]
```diff
@@ -181,14 +215,14 @@ struct BadImplStripper {
...
-    fn keep_impl(&self, ty: &Type) -> bool {
+    fn keep_impl(&self, ty: &Type, is_deref: bool) -> bool {
         if let Generic(_) = ty {
             // keep impls made on generics
             true
         } else if let Some(prim) = ty.primitive_type() {
             self.prims.contains(&prim)
         } else if let Some(did) = ty.def_id_no_primitives() {
-            self.keep_impl_with_def_id(did.into())
+            is_deref || self.keep_impl_with_def_id(did.into())
```
[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/passes/collect_trait_impls.rs:225]: Ok. I'm not quite sure I understand the logic but given that the test cases pass I'm ok with leaving it as is :)

**Comment by jyn514** [comment]
Oh oops, I forgot I asked you to add a couple comments - r=me with that done.

@bors r-

File: src/test/rustdoc/recursive-deref.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/test/rustdoc/recursive-deref.rs:73]: Sure.

**Comment by camelid** [comment]
@jyn514 why did you beta-nominate this? The previous attempt at this change had to be reverted, so it seems risky to backport this PR. Also, the regression it fixes is already on stable.

**Comment by jyn514** [comment]
@camelid because they're pretty bad regressions; several of them were announced in release notes and then missing from the next release. I'm ok with not backporting if you think it's risky, but I don't think it should be because the regressions are on stable.

**Comment by camelid** [comment]
I'd rather not backport since this PR could very possibly cause yet another regression, which we'd then have to revert on beta. Having a known regression on stable for one extra release seems better to me than the chance of accidentally introducing an unknown regression on beta right before release.

**Comment by camelid** [comment]
cc @rust-lang/rustdoc for backport discussion

**Comment by GuillaumeGomez** [comment]
Since I'm the author of this PR, I'll remain neutral. :)

**Comment by GuillaumeGomez** [comment]
The perf regression was expected. Maybe not this big though...

**Comment by jyn514** [comment]
@GuillaumeGomez What regression are you referring to?

**Comment by camelid** [comment]
@jyn514 I assume [this one](https://github.com/rust-lang/rust/pull/90422#issuecomment-955611455), and yeah, that is a pretty severe regression. Any idea if it can be reduced or does this change necessitate it?

**Comment by GuillaumeGomez** [comment]
We have to allocate a `HashSet` to ensure that we don't recurse infinitely and we also render a lot more of items. So it's most likely unavoidable. I have an idea which could reduce it a bit: currently we fill a `HashMap` with deref links and we also have a `HashSet` which checks that we don't recurse infinitely. Maybe we could use the `HashMap` directly.

When I tested it, I encountered on issue not on the fill part but on the render part. Might be worth taking another look at it.

**Comment by camelid** [comment]
>  I have an idea which could reduce it a bit: currently we fill a `HashMap` with deref links and we also have a `HashSet` which checks that we don't recurse infinitely. Maybe we could use the `HashMap` directly.

That sounds potentially promising. Also, I'd suggest trying `BTreeMap` since IIRC it's often more performant with small keys like `DefId`.

**Comment by jsha** [comment]
I'm not fully caught up on the details here, but is it feasible to instead set a small constant limit on recursion depth? E.g. go at most 3 levels deep?

**Comment by GuillaumeGomez** [comment]
I guess we can, but do we want to?

**Comment by camelid** [comment]
> I'm not fully caught up on the details here, but is it feasible to instead set a small constant limit on recursion depth? E.g. go at most 3 levels deep?

That could potentially incur unnecessary recursion since the same Deref impl could be processed 3 times when it only needs to be processed once. Also, it might be confusing for users if a particular type has more than 3 levels of Deref indirection. Having that much indirection is already confusing, but it doesn't seem great for rustdoc to make it more confusing by stopping early.

**Comment by jyn514** [comment]
Has anyone actually profiled the difference to see what's taking longer?

**Comment by GuillaumeGomez** [comment]
I didn't at least.

**Comment by apiraino** [comment]
Backport suspended, waiting for #90489 (see [Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202021-11-04.20.2354818/near/260278752))

**Comment by GuillaumeGomez** [comment]
cc @rust-lang/rustdoc

**Comment by jyn514** [comment]
@apiraino what feedback is needed? I left some comments in the Zulip thread already last week.

**Comment by apiraino** [comment]
Ok, let me [link your comments on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202021-11-04.20.2354818/near/260332428) here for reference.

Also, [discussed today on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202021-11-11.20.2354818/near/261146824)

**Comment by jyn514** [comment]
The conclusion in the compiler meeting was "we don't have enough background and this only touches rustdoc, so we defer to T-rustdoc's decision."

**Comment by apiraino** [comment]
Beta backport declined as per latest discussion, seems the aggreement is to let it flow to next beta (which should be tagged in about two weeks)

@rustbot label -beta-nominated

**Pull request merged**